app/src/lib.rs

```diff
@@ -493,6 +511,78 @@ impl PortalApp {
         self.router.remove_relay(url).await?;
         Ok(())
     }
+
+    pub async fn listen_invoice_requests(
+        &self,
+        evt: Arc<dyn InvoiceRequestListener>,
+    ) -> Result<(), AppError> {
+        let inner = InvoiceReceiverConversation::new(self.router.keypair().public_key());
+        let mut rx = self
+            .router
+            .add_and_subscribe(MultiKeyListenerAdapter::new(
+                inner,
+                self.router.keypair().subkey_proof().cloned(),
+            ))
+            .await?;
+
+        while let Ok(request) = rx.next().await.ok_or(AppError::ListenerDisconnected)? {
+            log::debug!("Received invoice request payment: {:?}", request);
+
+            let recipient: nostr::key::PublicKey = request.key.into();
+
+            let invoice = evt.on_invoice_requests(request.clone()).await?;
+
+            let invoice_response = InvoiceResponse {
+                request: request,
+                invoice: invoice.invoice,
+                payment_hash: invoice.payment_hash,
+            };
+
+            let conv = InvoiceSenderConversation::new(
+                invoice_response,
+                self.router.keypair().public_key(),
+                recipient,
+            );
+
+            self.router
+                .add_conversation(Box::new(OneShotSenderAdapter::new_with_user(
+                    recipient.to_owned(),
+                    vec![],
+                    conv,
+                )))
+                .await?;
+        }
+
+        Ok(())
```

Comment:
Now that you have rebased you should update this to match the other listeners (use the `self.runtime.add_task`) so that we can manage multiple parallel requests